gh-93065: Fix HAMT to iterate correctly over 7-level deep trees

1st1 opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->

vstinner commented on Include/internal/pycore_hamt.h:
```diff
@@ -5,7 +5,7 @@
 #endif
 
-#define _Py_HAMT_MAX_TREE_DEPTH 7
+#define _Py_HAMT_MAX_TREE_DEPTH 8
```
It lacks an important comment like: "8 should be enough for everyone", famous last words.

1st1 replied:
good point actually, i added an explainer

molaxx committed:
gh-93065: Fix HAMT to iterate correctly over 7-level deep trees.

Also while there, clarify a few things about why we reduce
the hash to 32 bits.

Co-authored-by: Yury Selivanov <yury@edgedb.com>

vstinner commented on Include/internal/pycore_hamt.h:
```diff
@@ -5,7 +5,19 @@
-#define _Py_HAMT_MAX_TREE_DEPTH 7
+
+/*
+HAMT tree is shaped by hashes of keys. Every group of 5 bits of a hash denotes
```
You may mention "gh-93065" somewhere for details.

ambv replied:
Nah, the current comment describes the design exactly as intended. The fact that there was a bug before is kind of spurious.

vstinner reviewed (approved):
LGTM. Thanks for adding a comment explaining the 8 constant!

ambv commented on Misc/NEWS.d/next/Core and Builtins/2022-05-21-23-21-37.gh-issue-93065.5I18WC.rst:
```diff
@@ -0,0 +1,5 @@
+
+The bug was discovered and fixed by Eli Libman. See [1] for more details.
+
+[1] https://github.com/MagicStack/immutables/issues/84
```
This doesn't look great in the actual change log later:
<img width="814" alt="Screen Shot 2022-05-23 at 19 38 32" src="https://user-images.githubusercontent.com/55281/169881776-93025663-503c-495f-afb2-1e46279db242.png">

I'll change it to an inline link and land.

ambv replied:
After:
<img width="797" alt="Screen Shot 2022-05-23 at 20 19 33" src="https://user-images.githubusercontent.com/55281/169882554-09403f96-eb62-4b73-bf54-3fe5fe74bf61.png">

ambv committed:
Change NEWS entry link to an inline link.
